# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxUserFontSet.cpp
# Commit: 5a39eb924188
# Full Hash: 5a39eb924188e504dad274a2148543c8d51dbae8
# Author: Andrew Osmond <aosmond@mozilla.com>
# Date: 2022-06-29 03:44:30
# Regressor Bug: 1771493
# File Overlap Count: 1
# Description:
#   Bug 1771493 - Part 2. Merge FontFaceSetImpl and gfxUserFontSet into a single class. r=emilio
#   
#   This patch merges FontFaceSetImpl and gfxUserFontSet into the same
#   class. This reduces the level of indirection and in theory makes it
#   easier to manage future threading issues.
# ==============================================================================

diff -r 3d30a8d83d77 -r 5a39eb924188 gfx/thebes/gfxUserFontSet.cpp
--- a/gfx/thebes/gfxUserFontSet.cpp	Tue Jun 28 21:46:59 2022 +0000
+++ b/gfx/thebes/gfxUserFontSet.cpp	Tue Jun 28 21:47:00 2022 +0000
@@ -254,11 +254,11 @@
   MOZ_ASSERT(NS_IsMainThread());
 }
 
-gfxFontSrcPrincipal* gfxFontFaceSrc::LoadPrincipal(
+already_AddRefed<gfxFontSrcPrincipal> gfxFontFaceSrc::LoadPrincipal(
     const gfxUserFontSet& aFontSet) const {
   MOZ_ASSERT(mSourceType == eSourceType_URL);
   if (mUseOriginPrincipal && mOriginPrincipal) {
-    return mOriginPrincipal;
+    return RefPtr{mOriginPrincipal}.forget();
   }
   return aFontSet.GetStandardFontLoadPrincipal();
 }
@@ -933,11 +933,15 @@
   }
 }
 
-gfxUserFontSet::~gfxUserFontSet() {
+gfxUserFontSet::~gfxUserFontSet() { Destroy(); }
+
+void gfxUserFontSet::Destroy() {
   gfxPlatformFontList* fp = gfxPlatformFontList::PlatformFontList();
   if (fp) {
     fp->RemoveUserFontSet(this);
   }
+
+  mFontFamilies.Clear();
 }
 
 already_AddRefed<gfxUserFontEntry> gfxUserFontSet::FindOrCreateUserFontEntry(
@@ -1259,7 +1263,7 @@
   }
 
   // Ignore principal when looking up a data: URI.
-  gfxFontSrcPrincipal* principal =
+  RefPtr<gfxFontSrcPrincipal> principal =
       IgnorePrincipal(aSrc.mURI) ? nullptr
                                  : aSrc.LoadPrincipal(*aUserFontEntry.mFontSet);
 